### PR TITLE
DatabaseTarget - Removed logEvent-parameter from CreateDatabaseParameter

### PR DESCRIPTION
**Pull request by snakefoot**
Update the new NLog 4.6 method CreateDatabaseParameter, so it is open for batching support.

### CODE REVIEW

**Comment by snakefoot** [comment]
Custom pre-release-build for those who wants to test (Just remove zip-extension):

[NLog.4.6.0-PR3131.nupkg.zip](https://github.com/NLog/NLog/files/2886704/NLog.4.6.0-PR3131.nupkg.zip)

**Comment by snakefoot** [comment]
Custom pre-release-build for those who wants to test performance (Just remove zip-extension):

[NLog.4.6.0-PR3131.nupkg.zip](https://github.com/NLog/NLog/files/2886733/NLog.4.6.0-PR3131.nupkg.zip)

File: src/NLog/Targets/DatabaseTarget.cs [hunk 1]
```diff
@@ -533,7 +539,18 @@ protected override void Write(LogEventInfo logEvent)
         {
             try
             {
-                WriteEventToDatabase(logEvent);
+                var connectionString = BuildConnectionString(logEvent);
+
+                //Always suppress transaction so that the caller does not rollback logging if they are rolling back their transaction.
+                using (TransactionScope transactionScope = new TransactionScope(TransactionScopeOption.Suppress))
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseTarget.cs]: this is another change then the PR suggests?

Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: Nope just me moving code around to optimize the batching-logic.

**Comment by 304NotModified** [comment]
Thanks! moved it to https://github.com/NLog/NLog/pull/3132, so we could easy integrate it into dev and master.

File: src/NLog/Targets/DatabaseTarget.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: OK thanks!

**Comment by snakefoot** [comment]
Could this become NLog 4.6, because this is actually a breaking change if NLog 4.6 is released.

**Comment by snakefoot** [comment]
Also ready to separate PR that just includes the breaking change in `CreateDatabaseParameter`. Just to prepare for this PR.

**Comment by 304NotModified** [comment]
please rebase on dev, thanks!

**Comment by snakefoot** [comment]
> @304NotModified  please rebase on dev, thanks!

Done

**Comment by D9001** [comment]
> Custom pre-release-build for those who wants to test performance (Just remove zip-extension):
> 
> [NLog.4.6.0-PR3131.nupkg.zip](https://github.com/NLog/NLog/files/2886733/NLog.4.6.0-PR3131.nupkg.zip)

> 2019-02-21 18:33:54.0713 Error DatabaseTarget(Name=wrapped_db): Error when writing to database. Exception: System.InvalidOperationException: SqlCommand.Prepare method requires all parameters to have a
n explicitly set type.
   at System.Data.SqlClient.SqlParameter.Prepare(SqlCommand cmd)
   at System.Data.SqlClient.SqlCommand.Prepare()
   at NLog.Targets.DatabaseTarget.WriteEventToDatabase(IDbCommand command, LogEventInfo logEvent, Layout commandLayout, IList`1 databaseParameterInfos, Boolean prepareDbCommand)
   at NLog.Targets.DatabaseTarget.Write(IList`1 logEvents)

I received this error. I'm going home tonight and will see if I'm missing an attribute on the parameter nodes, but I've essentially mirrored your example + added my connection string.

**Comment by snakefoot** [comment]
> I received this error. I'm going home tonight and will see if I'm missing an attribute on the parameter nodes, but I've essentially mirrored your example + added my connection string.

Thanks for testing. Have updated the xml in the original issue-post with dbType for each parameter. Please update your test config and try again. See also: https://docs.microsoft.com/en-us/dotnet/api/system.data.dbtype

**Comment by D9001** [comment]
> > I received this error. I'm going home tonight and will see if I'm missing an attribute on the parameter nodes, but I've essentially mirrored your example + added my connection string.
> 
> Thanks for testing. Have updated the xml in the original issue-post with dbType for each parameter. Please update your test config and try again. See also: https://docs.microsoft.com/en-us/dotnet/api/system.data.dbtype

I tried adding dbType earlier, but I used "int32" as the documentation for DatabaseTarget had an example in lower case. But maybe it's case sensitive. I will try it out with "Int32" instead.

**Comment by snakefoot** [comment]
> I tried adding dbType earlier, but I used "int32" as the documentation for DatabaseTarget had an example in lower case. But maybe it's case sensitive. I will try it out with "Int32" instead.

Else try doing this instead:

`<parameter name="@level" layout="${level:format=ordinal}" dbType="SqlDbType.Int" />`

**Comment by 304NotModified** [comment]
> System.InvalidOperationException: SqlCommand.Prepare method requires all parameters to have a
> n explicitly set type.

PS: I think we won't do that by default, nor we can do because of backwards comp. So specify al DBtype attributes indeed

**Comment by D9001** [comment]
> > I tried adding dbType earlier, but I used "int32" as the documentation for DatabaseTarget had an example in lower case. But maybe it's case sensitive. I will try it out with "Int32" instead.
> 
> Else try doing this instead:
> 
> `<parameter name="@level" layout="${level:format=ordinal}" dbType="SqlDbType.Int" />`

I tried both Int32 and SqlDbType.Int with no luck yet; the same error appears. Does this configuration appear to be correct?

```
  <targets>
    <target name="asyncDatabase" type="AsyncWrapper" timeToSleepBetweenBatches="10000" queueLimit="100" overflowAction="Grow">
      <target name="wrapped_db" connectionstring="${gdc:item=ConnectionString}" xsi:type="Database" prepareDbCommand="True" keepConnection="true">
        <commandText>
          INSERT INTO Logging ([Identifier]) VALUES (@Id);
        </commandText>
        <parameter name="@Id" dbType="SqlDbType.Int" layout="${event-properties:item=Id}"/>
      </target>
    </target>
  </targets>
```

**Comment by 304NotModified** [comment]
config looks valid. Could you please capture and post the internal log ? (trace level)

**Comment by D9001** [comment]
> config looks valid. Could you please capture and post the internal log ? (trace level)

I forgot to post an update earlier, but this was due to a brain fart where I hadn't cleaned/rebuilt the solution so the nlog.config file being used was from my first attempt. I was able to bind the parameter in my example as well as dbTypes of SqlDbType.VarChar by specifying the size/length attributes as well.

**Comment by 304NotModified** [comment]
🎉 great news! Looking forward to the results!

**Comment by 304NotModified** [comment]
@kwaak maybe this feature is also nice for you :)

**Comment by snakefoot** [comment]
@304NotModified Improved the logic when using non-string DbType where Layout-result is string.Empty. Returning default-value instead of errors about unable to convert empty string.

File: src/NLog/Targets/DatabaseTarget.cs [hunk 2]
```diff
@@ -629,32 +664,46 @@ protected override void Write(IList<AsyncLogEventInfo> logEvents)
...
+            }
 
-                //not really needed as there is no transaction at all.
-                transactionScope.Complete();
+            if (!prepareDbCommand || command.CommandText != commandText)
+            {
+                command.CommandText = commandText;
+                if (prepareDbCommand)
+                {
+                    command.Prepare();
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseTarget.cs]: I'm not sure if we should prepare every sql query (worst case).

I would expect this in the init method.

Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: Could make a cache of prepared DbCommands, and perform lookup using the key of ConnectionString + CommandText.

**Comment by 304NotModified** [comment]
PS ;)

>  would recommend against calling SqlCommand.Prepare() because the only potential benefit is saving network packets while the downside is taking up more server memory. 

https://dba.stackexchange.com/a/129721/32229

... to measure is to know... :)

**Comment by snakefoot** [comment]
> ... to measure is to know... :)

Don't mind removing the Prepare, and just change the parameter to be `ReuseDbCommand`

Extra caching is needed for SqlClient as SqlCommand-constructor starts as unprepared.

**Comment by snakefoot** [comment]
@D9001 Attached an updated build with caching of prepared DbCommands (See original post where new attachment has been added **NLog.4.6.0-PS3131.nupkg.zip**)

**Comment by snakefoot** [comment]
@D9001 Btw NLog tries to perform correct type-conversion based on DbType. But when using "special" types like  `SqlDbType.Int` then it fails to guess the correct value-type. You can help NLog by doing this (Adding parameterType="System.Int32"):

`<parameter name="@ManagerId" layout="${event-properties:item=ManagerId}" dbType="SqlDbType.Int" parameterType="System.Int32" />`

NLog will only guess correctly for "special" DbTypes when having these names: Date, TimeStamp, Double, Decimal, Bool, Guid

**Comment by D9001** [comment]
@snakefoot Thanks for the update. I've added the parameterType attributes for each parameter and have updated the NuGet package used in our test environment.

**Comment by snakefoot** [comment]
@D9001 Thank you for the update about what happens on the Server-side. Could you activate the NLog InternalLogger on Trace-Level, so one could see what happens on the Client-side?

See also: https://github.com/NLog/NLog/wiki/Internal-logging (Might have to check the log-file for sensitive information before posting to github).

Btw. if performance is important then you should NOT use the BufferingWrapper. Instead use the AsyncWrapper as shown in the first post for this PR. Make sure that your logger-rules points to the wrapper-target and not the database-target or else the batching will not be activated.

**Comment by D9001** [comment]
> @D9001 Thank you for the update about what happens on the Server-side. Could you activate the NLog InternalLogger on Trace-Level, so one could see what happens on the Client-side?
> 
> See also: https://github.com/NLog/NLog/wiki/Internal-logging (You might have to check the log-file for sensitive information before posting it here on github).
> 
> Btw. if performance is important then you should NOT use the BufferingWrapper (Mostly for throttle and to ensure high latency). Instead use the AsyncWrapper as shown in the first post for this PR. Make sure that your logger-rules points to the name of the wrapper-target and not the database-target or else the batching will not be activated.

Ah, thanks for explaining that. I thought I had read that BufferingWrapper provided the same functionality as AsyncWrapper as long as sliding timeouts were enabled. I figured it was better since I originally thought it would create one long insert statement. I'll swap to the AsyncWrapper and test that out with internal logging enabled also.

**Comment by D9001** [comment]
Here's the internal logging from a regular run (not performance) with some sensitive information I've omitted in between: https://justpaste.it/20eml

**Comment by snakefoot** [comment]
@D9001 That doesn't look good. Huge latency in the communication with the Database-instance. Guess a solution with sending an entire DataTable is needed for this case.

**Comment by snakefoot** [comment]
@D9001 Would you mind trying out #3150 ? (Use `xsi:type="DataTable"` instead of `xsi:type="Database" and configure the parameter `updateBatchSize="200"`)

**Comment by snakefoot** [comment]
@304NotModified The changes in this PR should still reduce client-side overhead for those using the DatabaseTarget with a local database-instance (without high latency)

**Comment by 304NotModified** [comment]
This can be closed in favor of #3150?

**Comment by snakefoot** [comment]
@304NotModified Have now removed the PrepareDbCommand-logic and instead just made the change to the new NLog 4.6 method: CreateDatabaseParameter

File: src/NLog/Targets/DatabaseTarget.cs [hunk 3]
```diff
@@ -884,7 +878,7 @@ protected virtual IDbDataParameter CreateDatabaseParameter(IDbCommand command, D
         /// </summary>
         /// <param name="logEvent">Current logevent.</param>
         /// <param name="parameterInfo">Parameter configuration info.</param>
-        protected internal virtual object GetParameterValue(LogEventInfo logEvent, DatabaseParameterInfo parameterInfo)
+        protected internal virtual object CreateDatabaseParameterValue(LogEventInfo logEvent, DatabaseParameterInfo parameterInfo)
```
Comment (304NotModified) [inline src/NLog/Targets/DatabaseTarget.cs]: IMO this is more a retrieve/get then a create. It won't create (`new`) things?

**Comment by snakefoot** [comment]
It will create a value for all situations except when extracted rawvalue matches dbtype


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


@304NotModified commented on this pull request.

________________________________

In src/NLog/Targets/DatabaseTarget.cs<https://github.com/NLog/NLog/pull/3131#discussion_r262997383>:

> @@ -884,7 +878,7 @@ protected virtual IDbDataParameter CreateDatabaseParameter(IDbCommand command, D
         /// </summary>
         /// <param name="logEvent">Current logevent.</param>
         /// <param name="parameterInfo">Parameter configuration info.</param>
-        protected internal virtual object GetParameterValue(LogEventInfo logEvent, DatabaseParameterInfo parameterInfo)
+        protected internal virtual object CreateDatabaseParameterValue(LogEventInfo logEvent, DatabaseParameterInfo parameterInfo)


IMO this is more a retrieve/get then a create. It won't create (new) things?

File: src/NLog/Targets/DatabaseTarget.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: It will create a value for all situations except when extracted rawvalue matches dbtype

Comment (304NotModified) [reply src/NLog/Targets/DatabaseTarget.cs]: It's a getOrCreate ;)

Not sure if we should have a virtual getOrCreate (need to check that)

IMO a create method should never "get”, so "get" is s better name (I'm trying to keep it easy to implement this method by others as it's virtual)

Comment (snakefoot) [reply src/NLog/Targets/DatabaseTarget.cs]: Renamed to GetDatabaseParameterValue

**Pull request merged**